### PR TITLE
Add ability to unpack scale resource packages and unbundle with flat structure

### PR DESCRIPTION
**Pull request by stephenk-msft**
-sa option skips all applicability checks and unpacks ALL packages, including scale resource packages, from the bundle
-pfn-flat option unbundles with a "flat" structure, meaning the unpacked packages are located side-by-side with the bundle folder. By default, the packages are unpacked to folders nested within the bundle's folder and are side-by-side with the bundle's footprint files. 

  | bundle folder name | Package folder   name | Languages   unpacked? | Scales unpacked? | Comments
-- | -- | -- | -- | -- | --
no options | <-d> | <-d>\<package   pfn> | No | No |  
-pfn | <-d>\<bundle   pfn> | <-d>\<bundle   pfn>\<package pfn> | No | No |  
-pfn -sl | <-d>\<bundle   pfn> | <-d>\<bundle   pfn>\<package pfn> | Yes | No |  
-pfn -sa | <-d>\<bundle   pfn> | <-d>\<bundle   pfn>\<package pfn> | Yes | Yes |  
-pfn-flat | <-d>\<bundle   pfn> | <-d>\<package   pfn> | No | No |  
-pfn-flat -sl | <-d>\<bundle   pfn> | <-d>\<package   pfn> | Yes | No |  
-pfn-flat -sa | <-d>\<bundle   pfn> | <-d>\<package   pfn> | Yes | Yes |  
-sl | <-d> | <-d>\<package   pfn> | Yes | No |  
-sa | <-d> | <-d>\<package   pfn> | Yes | Yes |  
-sa -sl | <-d> | <-d>\<package   pfn> | Yes | Yes | -sa includes -sl,   so we assume all packages should be unpacked

### CODE REVIEW

File: src/inc/Applicability.hpp [hunk 1]
```diff
@@ -53,8 +53,7 @@ namespace MSIX {
         void InitializeLanguages();
         void InitializeLanguages(IMsixApplicabilityLanguagesEnumerator* languagesEnumerator);
 
-        void AddPackageIfApplicable(ComPtr<IAppxPackageReader>& reader, std::string& packageName, 
-            const std::vector<Bcp47Tag>& packageLanguages, APPX_BUNDLE_PAYLOAD_PACKAGE_TYPE packageType, bool hasQualifiedResources);
+		void AddPackageIfApplicable(ComPtr<IAppxPackageReader>& reader, APPX_BUNDLE_PAYLOAD_PACKAGE_TYPE packageType, ComPtr<IAppxBundleManifestPackageInfo> bundleInfo);
```
Comment (JohnMcPMS) [inline src/inc/Applicability.hpp]: >	 [](start = 0, length = 1)

Tabs

Comment (JohnMcPMS) [reply src/inc/Applicability.hpp]: And throughout

---
In reply to: [288745380](https://github.com/Microsoft/msix-packaging/pull/146#discussion_r288745380) [](ancestors = 288745380)

File: src/msix/AppxBundleManifest.cpp [hunk 2]
```diff
@@ -141,8 +141,9 @@ namespace MSIX {
         const std::string& architecture,
         const std::string& publisher,
         std::vector<Bcp47Tag>& languages,
+		std::vector<std::string>& scales,
         APPX_BUNDLE_PAYLOAD_PACKAGE_TYPE packageType):
-        m_factory(factory), m_fileName(name), m_size(size), m_offset(offset), m_languages(std::move(languages)), m_packageType(packageType)
+        m_factory(factory), m_fileName(name), m_size(size), m_offset(offset), m_languages(std::move(languages)), m_scales(std::move(scales)), m_packageType(packageType)
```
Comment (JohnMcPMS) [inline src/msix/AppxBundleManifest.cpp]: >std::move [](start = 122, length = 9)

Don't move a &; if you want to move pass in a &&

Comment (JohnMcPMS) [reply src/msix/AppxBundleManifest.cpp]: And I see that languages breaks that, but I didn't review that apparently...

---
In reply to: [288747023](https://github.com/Microsoft/msix-packaging/pull/146#discussion_r288747023) [](ancestors = 288747023)

File: src/msix/AppxPackageObject.cpp [hunk 3]
```diff
@@ -29,6 +29,7 @@
 #include <limits>
 #include <algorithm>
 #include <array>
+#include <atlstr.h>
```
Comment (JohnMcPMS) [inline src/msix/AppxPackageObject.cpp]: I have my doubts that we have this available on linux/android/macos...

File: src/msix/AppxPackageObject.cpp [hunk 4]
```diff
@@ -245,16 +246,21 @@ namespace MSIX {
 
             for (const auto& package : bundleInfo->GetPackages())
             {
-                auto bundleInfoInternal = package.As<IAppxBundleManifestPackageInfoInternal>();
-                auto packageName = bundleInfoInternal->GetFileName();
-                auto packageStream = m_container->GetFile(Encoding::EncodeFileName(packageName));
+				LPWSTR packageName = nullptr;
+				auto bundleInfoInternal = package.As<IAppxBundleManifestPackageInfo>();
+				bundleInfoInternal->GetFileName(&packageName);
+				std::string packageNameStr = CW2A(packageName);
```
Comment (JohnMcPMS) [inline src/msix/AppxPackageObject.cpp]: >CW2A [](start = 33, length = 4)

If this is why you include atlstr.h, see unicodeconversion.hpp

File: src/msix/AppxPackageObject.cpp [hunk 5]
```diff
@@ -245,16 +246,21 @@ namespace MSIX {
 
             for (const auto& package : bundleInfo->GetPackages())
             {
-                auto bundleInfoInternal = package.As<IAppxBundleManifestPackageInfoInternal>();
-                auto packageName = bundleInfoInternal->GetFileName();
-                auto packageStream = m_container->GetFile(Encoding::EncodeFileName(packageName));
+				LPWSTR packageName = nullptr;
+				auto bundleInfoInternal = package.As<IAppxBundleManifestPackageInfo>();
+				bundleInfoInternal->GetFileName(&packageName);
```
Comment (JohnMcPMS) [inline src/msix/AppxPackageObject.cpp]: >GetFileName [](start = 24, length = 11)

Need to check the HRESULT

File: src/msix/AppxPackageObject.cpp [hunk 6]
```diff
@@ -245,16 +246,21 @@ namespace MSIX {
 
             for (const auto& package : bundleInfo->GetPackages())
             {
-                auto bundleInfoInternal = package.As<IAppxBundleManifestPackageInfoInternal>();
-                auto packageName = bundleInfoInternal->GetFileName();
-                auto packageStream = m_container->GetFile(Encoding::EncodeFileName(packageName));
+				LPWSTR packageName = nullptr;
+				auto bundleInfoInternal = package.As<IAppxBundleManifestPackageInfo>();
```
Comment (JohnMcPMS) [inline src/msix/AppxPackageObject.cpp]: >IAppxBundleManifestPackageInfo [](start = 41, length = 30)

Why did you change to this interface?  Seems like it just made things harder here.

Comment (JohnMcPMS) [inline src/msix/AppxPackageObject.cpp]: >bundleInfoInternal [](start = 9, length = 18)

Not internal any more.

Comment (JohnMcPMS) [reply src/msix/AppxPackageObject.cpp]: My suggestion would be to leave this all alone and simply .As it right before you need it below.

---
In reply to: [288749576](https://github.com/Microsoft/msix-packaging/pull/146#discussion_r288749576) [](ancestors = 288749576)

File: src/inc/Applicability.hpp (hunk 1 above)
Comment (msftrubengu) [inline src/inc/Applicability.hpp]: >ComPtr<IAppxBundleManifestPackageInfo> bundleInfo [](start = 112, length = 49)

const ComPtr<IAppxBundleManfiestPackageInfo>&

File: src/inc/AppxBundleManifest.hpp [hunk 7]
```diff
@@ -26,7 +26,6 @@ class IBundleInfo : public IUnknown
 };
 MSIX_INTERFACE(IBundleInfo, 0xff82ffcd,0x747a,0x4df9,0x88,0x79,0x85,0x3a,0xb9,0xdd,0x15,0xa1);
 
-// {32e6fcf0-729b-401d-9dbc-f927b494f9af}
```
Comment (msftrubengu) [inline src/inc/AppxBundleManifest.hpp]: can you get back this comment?

File: src/makemsix/main.cpp [hunk 8]
```diff
@@ -71,6 +77,12 @@ struct State
         return true;
     }
 
+    bool SkipAll()
```
Comment (msftrubengu) [inline src/makemsix/main.cpp]: >SkipAll [](start = 9, length = 7)

This should be name more accordingly to applicability (aka SkipAllApplicability/SkipApplicability). SkipAll sounds, to me, that is a union of all the Skip* methods this struct has, which isn't true

File: src/msix/ApplicabilityCommon.cpp [hunk 9]
```diff
@@ -155,16 +157,29 @@ namespace MSIX {
...
-        const std::vector<Bcp47Tag>& packageLanguages, APPX_BUNDLE_PAYLOAD_PACKAGE_TYPE packageType, bool hasQualifiedResources)
+	void Applicability::AddPackageIfApplicable(ComPtr<IAppxPackageReader>& reader, APPX_BUNDLE_PAYLOAD_PACKAGE_TYPE packageType, ComPtr<IAppxBundleManifestPackageInfo> bundlePackageInfo)
     {
-        // If there are not qualified resources the package is always applicable
-        if (!hasQualifiedResources)
+	    auto bundlePackageInfoInternal = bundlePackageInfo.As<IAppxBundleManifestPackageInfoInternal>();
+	    auto packageName = bundlePackageInfoInternal->GetFileName();
+	    auto packageLanguages = bundlePackageInfoInternal->GetLanguages();
+	    auto packageScales = bundlePackageInfoInternal->GetScales();
+	    bool hasQualifiedResources = bundlePackageInfoInternal->HasQualifiedResources();
```
Comment (msftrubengu) [inline src/msix/ApplicabilityCommon.cpp]: >hasQualifiedResources [](start = 10, length = 21)

there's no need for this local, i would just call it in the if

File: src/inc/AppxBundleManifest.hpp [hunk 10]
```diff
@@ -143,6 +146,7 @@ namespace MSIX {
         std::uint64_t m_size;
         std::uint64_t m_offset;
         std::vector<Bcp47Tag> m_languages;
+        std::vector<std::string> m_scales;
```
Comment (msftrubengu) [inline src/inc/AppxBundleManifest.hpp]: >std::string [](start = 20, length = 11)

why a string? can it be just UINT32?

File: src/makemsix/main.cpp [hunk 11]
```diff
@@ -331,6 +343,10 @@ int main(int argc, char* argv[])
                     [](State& state, const std::string&) { return state.SkipLanguage(); }),
                 Option("-sp", false, "Only for bundles. Skips matching packages with of the same system. By default unpacked application packages will only match the platform.",
                     [](State& state, const std::string&) { return state.SkipPlatform(); }),
+                Option("-sa", false, "Only for bundles. Skips all applicability checks for packages.",
+                    [](State& state, const std::string&) { return state.SkipApplicability(); }),
+                Option("-flat", false, "Only for bundles. Places unpacked packages side-by-side with the bundle folder. By default unpacked packages will be nested inside the bundle folder",
+	                [](State& state, const std::string&) { return state.UnpackWithFlatStructure(); }),
```
Comment (JohnMcPMS) [inline src/makemsix/main.cpp]: >	 [](start = 0, length = 1)

Tab

Comment (JohnMcPMS) [reply src/makemsix/main.cpp]: Still more in your changes

---
In reply to: [288787944](https://github.com/Microsoft/msix-packaging/pull/146#discussion_r288787944) [](ancestors = 288787944)

**Comment by msftrubengu** [comment]
>     class AppxBundleQualifiedResource final : public MSIX::ComClass<AppxBundleQualifiedResource, IAppxManifestQualifiedResource, IAppxManifestQualifiedResourceUtf8>

Now that we have scales, this needs to be updated.

---
Refers to: src/inc/AppxBundleManifest.hpp:74 in ff5f209. [](commit_id = ff5f2090af884b2a9340df614acc011f5f186dd7, deletion_comment = False)

**Comment by msftrubengu** [comment]
>     HRESULT STDMETHODCALLTYPE AppxBundleManifestPackageInfo::GetResources(IAppxManifestQualifiedResourcesEnumerator **resources) noexcept try

this needs to be updated too

---
Refers to: src/msix/AppxBundleManifest.cpp:188 in ff5f209. [](commit_id = ff5f2090af884b2a9340df614acc011f5f186dd7, deletion_comment = False)

File: src/msix/IXml.cpp [hunk 12]
```diff
@@ -24,7 +24,7 @@ static const char* attributeNames[] = {
     /* Language                               */"Language",
     /* MinVersion                             */"MinVersion",
     /* Dependencies_Tdf_MaxVersionTested      */"MaxVersionTested",
-    /* Package_Applications_Application_Id    */"Id",
```
Comment (JohnMcPMS) [inline src/msix/IXml.cpp]: Bad merge of this; the value has been lost.  We should have a static_assert it would seem.

File: src/msix/AppxPackageObject.cpp [hunk 13]
```diff
@@ -331,9 +331,9 @@ namespace MSIX {
                 APPX_BUNDLE_PAYLOAD_PACKAGE_TYPE packageType;
                 ThrowHrIfFailed(package->GetPackageType(&packageType));
                 
+                auto bundlePackageInfo = package.As<IAppxBundleManifestPackageInfo>();
```
Comment (msftrubengu) [inline src/msix/AppxPackageObject.cpp]: >auto bundlePackageInfo = package.As<IAppxBundleManifestPackageInfo>(); [](start = 16, length = 70)

No need for this QI,  package is already an IAppxBundleManifestPackageInfo. you can rename it to make clear what it is

File: src/inc/AppxBundleManifest.hpp (hunk 10 above)
Comment (stephenk-msft) [reply src/inc/AppxBundleManifest.hpp]: That just happened to be the default value when getting the attribute value when calling resourceNode->GetAttributeValue(). 

I can change to UINT32.

File: src/msix/IXml.cpp (hunk 12 above)
Comment (stephenk-msft) [reply src/msix/IXml.cpp]: Should I put the value back in?

File: src/msix/ApplicabilityCommon.cpp [hunk 14]
```diff
@@ -155,16 +157,28 @@ namespace MSIX {
         }
     }
 
-    void Applicability::AddPackageIfApplicable(ComPtr<IAppxPackageReader>& reader, std::string& packageName, 
-        const std::vector<Bcp47Tag>& packageLanguages, APPX_BUNDLE_PAYLOAD_PACKAGE_TYPE packageType, bool hasQualifiedResources)
+	void Applicability::AddPackageIfApplicable(ComPtr<IAppxPackageReader>& reader, APPX_BUNDLE_PAYLOAD_PACKAGE_TYPE packageType, const ComPtr<IAppxBundleManifestPackageInfo>& bundlePackageInfo)
```
Comment (JohnMcPMS) [inline src/msix/ApplicabilityCommon.cpp]: >	 [](start = 0, length = 1)

Tab

File: src/msix/AppxBundleManifest.cpp [hunk 15]
```diff
@@ -187,14 +192,22 @@ namespace MSIX {
...
         {
             auto resource = ComPtr<IAppxManifestQualifiedResource>::Make<AppxBundleQualifiedResource>(m_factory, bcp47.GetFullTag());
-            languages.push_back(std::move(resource));
+            m_resources.push_back(std::move(resource));
         }
+
+        for (auto& scale : m_scales)
+        {
+	        std::string language;
+	        auto resource = ComPtr<IAppxManifestQualifiedResource>::Make<AppxBundleQualifiedResource>(m_factory, language, scale);
```
Comment (msftrubengu) [inline src/msix/AppxBundleManifest.cpp]: >Make<AppxBundleQualifiedResource>(m_factory, language, scale); [](start = 65, length = 62)

You don't need the empty string. Create a constructor that takes the factory and the scale.

File: src/msix/ApplicabilityCommon.cpp [hunk 16]
```diff
@@ -155,16 +157,28 @@ namespace MSIX {
...
 
-    void Applicability::AddPackageIfApplicable(ComPtr<IAppxPackageReader>& reader, std::string& packageName, 
-        const std::vector<Bcp47Tag>& packageLanguages, APPX_BUNDLE_PAYLOAD_PACKAGE_TYPE packageType, bool hasQualifiedResources)
+    void Applicability::AddPackageIfApplicable(ComPtr<IAppxPackageReader>& reader, APPX_BUNDLE_PAYLOAD_PACKAGE_TYPE packageType, const ComPtr<IAppxBundleManifestPackageInfo>& bundlePackageInfo)
     {
+        auto bundlePackageInfoInternal = bundlePackageInfo.As<IAppxBundleManifestPackageInfoInternal>();
+        auto packageName = bundlePackageInfoInternal->GetFileName();
+        auto packageLanguages = bundlePackageInfoInternal->GetLanguages();
+        auto packageScales = bundlePackageInfoInternal->GetScales();
+
```
Comment (msftrubengu) [inline src/msix/ApplicabilityCommon.cpp]: tabs

File: src/msix/AppxBundleManifest.cpp [hunk 17]
```diff
@@ -187,14 +192,21 @@ namespace MSIX {
...
         for(auto& bcp47 : m_languages)
         {
             auto resource = ComPtr<IAppxManifestQualifiedResource>::Make<AppxBundleQualifiedResource>(m_factory, bcp47.GetFullTag());
-            languages.push_back(std::move(resource));
+            m_resources.push_back(std::move(resource));
         }
+
+        for (auto& scale : m_scales)
+        {
+	        auto resource = ComPtr<IAppxManifestQualifiedResource>::Make<AppxBundleQualifiedResource>(m_factory, scale);
```
Comment (msftrubengu) [inline src/msix/AppxBundleManifest.cpp]: >	 [](start = 0, length = 1)

here too (and next line)
